packages/intrepid2/assembly-examples/StructuredAssembly.hpp

```diff
@@ -0,0 +1,241 @@
+  using IntegrationTools   = IntegrationTools<DeviceType>;
+  // dimensions of the returned view are (C,F,F)
+  
+  Intrepid2::ScalarView<Intrepid2::Orientation,DeviceType> orientations("orientations", geometry.numCells() );
+  geometry.orientations(orientations, 0, -1);
+  
+  shards::CellTopology cellTopo = geometry.cellTopology();
+  
+  auto basis1 = getBasis< BasisFamily >(cellTopo, fs1, polyOrder1);
+  auto basis2 = getBasis< BasisFamily >(cellTopo, fs2, polyOrder2);
+  
+  int numFields1 = basis1->getCardinality();
+  int numFields2 = basis2->getCardinality();
+  int numCells = geometry.numCells();
+    
+  // local stiffness matrix:
+  ScalarView<Scalar,DeviceType> cellStiffness("cell stiffness matrices",numCells,numFields1,numFields2);
+  ScalarView<Scalar,DeviceType> worksetCellStiffness("cell stiffness workset matrices",worksetSize,numFields1,numFields2);
+
+  auto cubature = DefaultCubatureFactory::create<DeviceType>(cellTopo,polyOrder1 + polyOrder2);
+  auto tensorCubatureWeights = cubature->allocateCubatureWeights();
+  TensorPoints<PointScalar,DeviceType> tensorCubaturePoints  = cubature->allocateCubaturePoints();
+  
+  cubature->getCubature(tensorCubaturePoints, tensorCubatureWeights);
+  
+  BasisValues<Scalar,DeviceType> basis1Values = basis1->allocateBasisValues(tensorCubaturePoints, op1);
+  basis1->getValues(basis1Values, tensorCubaturePoints, op1);
+  
+  BasisValues<Scalar,DeviceType> basis2Values = basis2->allocateBasisValues(tensorCubaturePoints, op2);
+  basis2->getValues(basis2Values, tensorCubaturePoints, op2);
+      
+  int cellOffset = 0;
+  
+  auto jacobianAndCellMeasureTimer = Teuchos::TimeMonitor::getNewTimer("Jacobians");
+  auto fstIntegrateCall = Teuchos::TimeMonitor::getNewTimer("transform + integrate()");
+  
+  Data<PointScalar,DeviceType> jacobian       = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize);
+  Data<PointScalar,DeviceType> jacobianDividedByJacobianDet = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize); // jacobianDividedByJacobianDet has same underlying structure as jacobian
+  Data<PointScalar,DeviceType> jacobianDet    = CellTools<DeviceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,DeviceType> jacobianDetInv = CellTools<DeviceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,DeviceType> jacobianInv    = CellTools<DeviceType>::allocateJacobianInv(jacobian);
+  TensorData<PointScalar,DeviceType> cellMeasures = geometry.allocateCellMeasure(jacobianDet, tensorCubatureWeights);
+  
+  // lazily-evaluated transformed basis values (temporary to allow integralData allocation)
+  auto transformedBasis1ValuesTemp = transform(basis1Values, fs1, op1, jacobian, jacobianDet, jacobianInv, jacobianDetInv, jacobianDividedByJacobianDet);
+  auto transformedBasis2ValuesTemp = transform(basis2Values, fs2, op2, jacobian, jacobianDet, jacobianInv, jacobianDetInv, jacobianDividedByJacobianDet);
```

Comment:
Being able to do structured integration by passing the differential operators `op1` and `op2` was something I didn't realize would be possible, but that's really useful for the matrix-free work I'm doing!

I think these sort of interfaces will also be useful for integrating structured and matrix-free capabilities with other packages like Panzer.